### PR TITLE
#107 Setup Danger

### PR DESCRIPTION
**Pull request by kevnm67**
## Pull request checklist

- [x] All tests pass. 
- [x] Demo project builds and runs.
- [x] I have resolved merge conflicts.
- [x] I have followed the [coding style](https://github.com/jessesquires/HowToContribute#style-guidelines) and [contributing guidelines](https://github.com/jessesquires/HowToContribute).

#### This fixes issue #
#107 Setup Danger

## What's in this pull request?
- [Added] Dangerfile with related files for installing required plugins.
- [Added] Codeowners file.
- [Updated] Travis config to setup danger and exec danger at the end of the script phase.

### CODE REVIEW

**Comment by kevnm67** [comment]
@jessesquires I have danger working locally and in the [PR I created from my forked version](https://github.com/kevnm67/JSQDataSourcesKit/pull/1#issuecomment-470643019).  A few things to note:

1.  The test suite had compiler errors.  It looks like it's been happening for a few months according to travis.  I fixed those in this branch.  Not sure if you want these broken out to a separate PR? or...

2.  After fixing the above I ran into 2 more issues: A failed test case and a range exception.  I fixed the range exception and will fix the failed test case (if you want).

Im going to assume you'd want those in a separate PR unless you say otherwise.

**Comment by jessesquires** [comment]
Hey @kevnm67 -- this is looking great! 💯 🎉 

Commented on #121. Sorry about that! That was from a recent commit. All issues on `develop` should be resolved now. 

I've also added `CODEOWNERS` on `develop` (to make this diff smaller).

Could you rebase/merge with `develop` and revert the code and swiftlint changes?

File: .ruby-version [hunk 1]
```diff
@@ -0,0 +1 @@
+2.5.1
```
Comment (jessesquires) [inline .ruby-version]: do we need this?

i think in `.travis.yml` we can just specify `rvm: 2.5.1`

(would rather keep this in one place)

File: .travis.yml [hunk 2]
```diff
@@ -1,5 +1,12 @@
-language: objective-c
+language: swift
```
Comment (jessesquires) [inline .travis.yml]: hm. does this work? travis-ci docs still say to use `objective-c` 😄

File: .travis.yml [hunk 3]
```diff
@@ -1,5 +1,12 @@
...
 osx_image: xcode10.1
+addons:
+  homebrew:
+    packages:
+    - swiftlint
+    update: true
```
Comment (jessesquires) [inline .travis.yml:8]: wow, this is new. 💯 TIL

File: .travis.yml [hunk 4]
```diff
@@ -22,9 +29,8 @@ env:
    - DESTINATION="OS=12.1,name=Apple TV" SDK="$TVOS_SDK" RUN_TESTS="YES" BUILD_EXAMPLE="NO" POD_LINT="NO"
 
 before_install:
-   - gem install cocoapods --pre --no-rdoc --no-ri --no-document --quiet
-   - brew update
-   - brew outdated swiftlint || brew upgrade swiftlint
+  - rvm install 2.5.1
```
Comment (jessesquires) [inline .travis.yml]: per comment above, we can remove this

File: .ruby-version (hunk 1 above)
Comment (jessesquires) [reply .ruby-version]: ah, looks like we don't need this:

https://docs.travis-ci.com/user/languages/ruby/#using-ruby-version

File: .travis.yml [hunk 5]
```diff
@@ -1,5 +1,12 @@
-language: objective-c
+language: swift
 osx_image: xcode10.1
+addons:
+  homebrew:
+    packages:
+    - swiftlint
+    update: true
+cache:
+  bundler: true
```
Comment (jessesquires) [inline .travis.yml]: would prefer using `cache: bundler`

File: .travis.yml [hunk 6]
```diff
@@ -22,9 +29,8 @@ env:
    - DESTINATION="OS=12.1,name=Apple TV" SDK="$TVOS_SDK" RUN_TESTS="YES" BUILD_EXAMPLE="NO" POD_LINT="NO"
 
 before_install:
-   - gem install cocoapods --pre --no-rdoc --no-ri --no-document --quiet
-   - brew update
-   - brew outdated swiftlint || brew upgrade swiftlint
+  - rvm install 2.5.1
+  - bundle install --without=documentation
```
Comment (jessesquires) [inline .travis.yml]: we don't need this. travis will call `bundle install` automatically

let's update `before_install:` to be:

```
before_install:
    - gem update --system
    - gem install bundler
```

(to install bundler 2.0 https://docs.travis-ci.com/user/languages/ruby/#bundler-20)

Comment (jessesquires) [inline .travis.yml]: for below, L40, let's `bundle exec` the `pod lib lint` line

File: .travis.yml [hunk 7]
```diff
@@ -48,6 +54,8 @@ script:
 - if [ $RUN_TESTS == "YES" ]; then
       xcodebuild test -project JSQDataSourcesKit.xcodeproj -scheme JSQDataSourcesKit -destination "platform=iOS Simulator,name=iPhone X" CODE_SIGNING_REQUIRED=NO;
   fi
+  
+- bundle exec danger
```
Comment (jessesquires) [inline .travis.yml:58]: 👏

File: Dangerfile [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+# if github.pr_title.include? "[WIP]"
+# 	auto_label.wip=(github.pr_json["number"])
+# else
+# 	auto_label.remove("WIP")
+# 	# If you want to delete label
+# 	# auto_label.delete("WIP")
+# end
+
+# Let people say that this isn't worth a CHANGELOG entry in the PR if they choose
+declared_trivial = (github.pr_title + github.pr_body).include?("#trivial") || !has_app_changes
```
Comment (jessesquires) [inline Dangerfile]: let's remove this

File: Dangerfile [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+
+# Changelog entries are required for changes to library files.
+no_changelog_entry = !git.modified_files.include?("CHANGELOG.md")
+
+if has_app_changes && no_changelog_entry && !declared_trivial
+  warn("Any changes to library code should be reflected in the Changelog. Please consider adding a note there and adhere to the [Changelog Guidelines](https://github.com/jessesquires/JSQDataSourcesKit/CHANGELOG.md).")
```
Comment (jessesquires) [inline Dangerfile]: here, let's say, "There's no changelog entry. Do you need to add one?"

File: Dangerfile [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+### Auto label
+###
+
+# if github.pr_title.include? "[WIP]"
+# 	auto_label.wip=(github.pr_json["number"])
+# else
+# 	auto_label.remove("WIP")
+# 	# If you want to delete label
+# 	# auto_label.delete("WIP")
+# end
```
Comment (jessesquires) [inline Dangerfile]: let's remove L35 to here 😄

File: Dangerfile [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+end
+
+### WARNINGS
+
+warn('Changes to CI/CD files') if didModify(@DANGER_CI_CD_FILES)
+warn('Changes to dependency related files') if didModify(@DANGER_DEPENDENCY_FILES)
+warn("Big PR, try to keep changes smaller if you can") if git.lines_of_code > 500
+
+# Make it more obvious that a PR is a work in progress and shouldn't be merged yet
+warn("PR is classed as Work in Progress") if github.pr_title.include? "[WIP]"
```
Comment (jessesquires) [inline Dangerfile]: let's remove this

File: Dangerfile [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+		end
+	end
+	return did_modify_files
+end
+
+### WARNINGS
+
+warn('Changes to CI/CD files') if didModify(@DANGER_CI_CD_FILES)
+warn('Changes to dependency related files') if didModify(@DANGER_DEPENDENCY_FILES)
+warn("Big PR, try to keep changes smaller if you can") if git.lines_of_code > 500
```
Comment (jessesquires) [inline Dangerfile]: let's say: "This is a large pull request! Can you break it up into multiple smaller ones instead?"

File: Dangerfile [hunk 13]
```diff
@@ -0,0 +1,82 @@
+
+has_app_changes = !git.added_files.grep(/Source.*\.swift/).empty?
```
Comment (jessesquires) [inline Dangerfile]: could we call this ~~`has_lib_changes`~~ `has_source_changes` ?

File: Dangerfile [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+modified_prose_lintable_files = (git.added_files.grep(%r{.*\.md/}) + git.modified_files.grep(%r{.*\.md/}))
+
+unless modified_prose_lintable_files.empty?
+  prose.lint_files modified_prose_lintable_files
+  prose.language = "en-us"
+  prose.ignored_words = ["JSQDataSourcesKit", "jessesquires", "enum", "enums", "CocoaPods"]
+  prose.ignore_acronyms = true
+  prose.ignore_numbers = true
+  prose.check_spelling modified_prose_lintable_files
+end
```
Comment (jessesquires) [inline Dangerfile:28]: nice! 👏

File: Dangerfile [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+if has_app_changes && no_changelog_entry && !declared_trivial
+  warn("Any changes to library code should be reflected in the Changelog. Please consider adding a note there and adhere to the [Changelog Guidelines](https://github.com/jessesquires/JSQDataSourcesKit/CHANGELOG.md).")
+end
+
+if has_app_changes && !has_test_changes
+	warn("Library files were updated without test coverage.  Consider updating or adding tests to cover changes.")
+end
+
+# Run danger-prose to lint and check spelling.
+modified_prose_lintable_files = (git.added_files.grep(%r{.*\.md/}) + git.modified_files.grep(%r{.*\.md/}))
```
Comment (jessesquires) [inline Dangerfile]: could we name this `markdown_files` ?

File: Dangerfile [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+# Run SwiftLint
+
+# Note: The argument "--force-exclude" is passed to SwiftLint by the danger-swiftlint plugin
+#       But this SwiftLint argument doesn't work properly & has issues.
+#       So instead, we re-disable it, and filter the list of files manually :-/
+files_to_lint = (git.modified_files - git.deleted_files) + git.added_files
+files_to_lint.reject! { |f| f.start_with?('scripts/', 'docs/') }
+swiftlint.lint_files(files_to_lint,
+	additional_swiftlint_args: '--no-force-exclude'
+)
```
Comment (jessesquires) [inline Dangerfile]: I think we can update all the swiftlint stuff to be:

```ruby
swiftlint.verbose = true
swiftlint.config_file = './.swiftlint.yml'
swiftlint.lint_files(inline_mode: true, fail_on_error: true)
```

File: Gemfile [hunk 17]
```diff
@@ -0,0 +1,9 @@
+source "https://rubygems.org"
+
+gem 'cocoapods'
+gem 'danger'
```
Comment (jessesquires) [inline Gemfile]: let's specify versions for these:

```spotify
gem 'cocoapods', '~> 1.6.0'
gem 'danger', '~> 5.0'
```

File: Gemfile [hunk 18]
```diff
@@ -0,0 +1,9 @@
...
+gem 'danger-swiftlint'
+gem 'danger-xcode_summary'
+gem 'xcpretty-json-formatter'
```
Comment (jessesquires) [inline Gemfile:8]: might as well explicitly specify `xcpretty` too

File: scripts/test_danger.sh [hunk 19]
```diff
@@ -0,0 +1,23 @@
...
+
+    # Use the provided PR to run danger against.
+    bundle exec danger pr https://github.com/kevnm67/JSQDataSourcesKit/pull/$(get_pr_value $1)
+}
+
+######################
+# Run danger locally #
+######################
+
+run_against_pr 1
```
Comment (jessesquires) [inline scripts/test_danger.sh]: nice! 👏

File: scripts/test_danger.sh [hunk 20]
```diff
@@ -0,0 +1,23 @@
...
+    else
+        echo "1"
+    fi
+}
+
+function run_against_pr {
+    cd ..; bundle install
+
+    # Use the provided PR to run danger against.
+    bundle exec danger pr https://github.com/kevnm67/JSQDataSourcesKit/pull/$(get_pr_value $1)
```
Comment (jessesquires) [inline scripts/test_danger.sh]: this should be `jessesquires` not `kevnm67`, right? 😄

File: scripts/test_danger.sh [hunk 21]
```diff
@@ -0,0 +1,23 @@
+#!/bin/bash
+
+# Get the PR number provided or set the default to 1. 
+function get_pr_value {
+    if [ $# = 1 ] ; then
+        echo $1
+    else
+        echo "1"
```
Comment (jessesquires) [inline scripts/test_danger.sh]: instead , maybe we could just error if you don't provide a PR number?

**Review by jessesquires — COMMENTED** [review]
Left a bunch of inline comments but overall this is awesome! 💯 🥇 

Thanks so much @kevnm67 !

File: scripts/test_danger.sh (hunk 21 above)
Comment (kevnm67) [reply scripts/test_danger.sh]: Sure, that makes sense.  Probably a better approach as well ;)

File: scripts/test_danger.sh (hunk 20 above)
Comment (kevnm67) [reply scripts/test_danger.sh]: :) Indeed... meant to switch this over after you added the danger token.  Ill fix this now.

File: .ruby-version (hunk 1 above)
Comment (kevnm67) [reply .ruby-version]: You can use either.  My reasoning: Ensure the ruby version is locked/consistent whether running on travis or locally.  

Which do you prefer?

File: .travis.yml (hunk 2 above)
Comment (kevnm67) [reply .travis.yml]: Actually, I don't think we need to specify the language at all. This works:  
```yaml
os: osx
osx_image: xcode10.1
```

File: Dangerfile (hunk 13 above)
Comment (kevnm67) [reply Dangerfile]: of course! Changed.

File: Dangerfile (hunk 12 above)
Comment (kevnm67) [reply Dangerfile]: Done.

File: Dangerfile (hunk 10 above)
Comment (kevnm67) [reply Dangerfile]: NOTE: Was commenting out bc I needed the label on this repo + the danger token.  If you want to auto-label PR's for certain keywords, let me know ;)
Removed.

File: Dangerfile (hunk 15 above)
Comment (kevnm67) [reply Dangerfile]: This was a guess as to what you'd want to lint using prose.  Is there anything else I should add? or exclude?

File: Dangerfile (hunk 16 above)
Comment (kevnm67) [reply Dangerfile]: Ok, ill try that 🤞.  I've had issues with two other projects and the way I wrote it here is what's been working (although not ideal, obviously).

Out of scope, but have you used [code climate](https://codeclimate.com/quality)?  I use it in multiple projects and it works great with swiftlint + inline comments.

File: Gemfile (hunk 17 above)
Comment (kevnm67) [reply Gemfile]: Good call. Done

File: scripts/test_danger.sh (hunk 19 above)
Comment (kevnm67) [reply scripts/test_danger.sh]: 😊

File: .travis.yml (hunk 2 above)
Comment (kevnm67) [reply .travis.yml]: > hm. does this work? travis-ci docs still say to use `objective-c` 😄

Magic :) I can't remember where/when I read about this setting but have used it for a year or so.  See my previous comment, though.

**Comment by kevnm67** [comment]
> Hey @kevnm67 -- this is looking great! 💯 🎉
> 
> Commented on #121. Sorry about that! That was from a recent commit. All issues on `develop` should be resolved now.
> 
> I've also added `CODEOWNERS` on `develop` (to make this diff smaller).
> 
> Could you rebase/merge with `develop` and revert the code and swiftlint changes?

Thanks!
No worries.  Thanks for adding the file in develop to ↓ the diff size ;)

Done.  The only reason I updated the swiftlint config was in response to swiftlint compiler warnings.  Something about not matching... reverted.  Weird, though.  Looks like swiftlint's repo using those rules 🤷‍♂️

**Comment by kevnm67** [comment]
@jessesquires Looks like travis is still missing the danger token environment variable:
```For your GitHub repo, you need to expose: DANGER_GITHUB_API_TOKEN```

Also, you may want to update your repo settings for travis and migrate to Travis to GitHub Apps now that [services has been deprecated](https://developer.github.com/changes/2018-04-25-github-services-deprecation/).

File: .ruby-version (hunk 1 above)
Comment (jessesquires) [reply .ruby-version]: let's go with specifying in only the `.travis.yml` file for now. if needed, we can revisit.

File: Dangerfile [hunk 22]
```diff
@@ -0,0 +1,59 @@
+
+has_source_changes = !git.added_files.grep(/Source.*\.swift/).empty?
```
Comment (jessesquires) [inline Dangerfile]: this should just be `/Source/`

File: Dangerfile [hunk 23]
```diff
@@ -0,0 +1,59 @@
...
+			did_modify_files = true
+		end
+	end
+	return did_modify_files
+end
+
+### WARNINGS
+
+warn('Changes to CI/CD files') if didModify(@DANGER_CI_CD_FILES)
+warn('Changes to dependency related files') if didModify(@DANGER_DEPENDENCY_FILES)
```
Comment (jessesquires) [inline Dangerfile]: Thinking about this more...

Actually, I don't think we need these specific warnings. These are rarely modified by contributors anyway, so I should always notice. If it ever becomes an issue, we can revisit.

**Comment by jessesquires** [comment]
@kevnm67 Thanks for making those changes! Left a couple more comments.

>  Looks like travis is still missing the danger token environment variable

Double-checked. It's definitely there. I think we might need to merge this first for it to register. After these last changes, let's get this merged, then we can do another PR for #122 (and maybe some other random PRs to test).

> Also, you may want to update your repo settings for travis and migrate to Travis to GitHub Apps now that services has been deprecated.

Ah! 💯 I'll look into this!

**Comment by kevnm67** [comment]
@jessesquires Made all the changes from your latest review ;)

**Comment by jessesquires** [comment]
Thanks @kevnm67 !

Just sent you an invite to be a collaborator on this repo 😄

**Comment by kevnm67** [comment]
Awesome!  Thanks @jessesquires, I appreciate it.

**Pull request merged**